File: src/main/java/fun/bookish/vertx/activemq/client/subscriber/ActiveMQSubscriberImpl.java
package fun.bookish.vertx.activemq.client.subscriber;

import fun.bookish.vertx.activemq.client.cache.ActiveMQCacheManager;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import org.apache.activemq.command.ActiveMQTextMessage;

import javax.jms.JMSException;
import javax.jms.Session;
import javax.jms.Topic;
import javax.jms.TopicSubscriber;
import java.util.concurrent.atomic.AtomicReference;


public class ActiveMQSubscriberImpl implements ActiveMQSubscriber {

    private final String key;
    private final Vertx vertx;
    private final ActiveMQCacheManager cacheManager;
    private final Session session;
    private final String destination;
    private Topic topic;

    public ActiveMQSubscriberImpl(String key,Vertx vertx,Session session,String destination, ActiveMQCacheManager cacheManager){
        this.key = key;
        this.vertx = vertx;
        this.session = session;
        this.destination = destination;
        this.cacheManager = cacheManager;
        try {
            this.topic = session.createTopic(destination);
        }catch (JMSException e){
            throw new IllegalArgumentException("ActiveMQ主题订阅者创建失败, topic = " + topic);
        }
    }

    private final AtomicReference<Future<TopicSubscriber>> subscriberRef = new AtomicReference<>();

    @Override
    public String getKey() {
        return this.key;
    }

    /**
     * 开启订阅监听
     */
    @SuppressWarnings("Duplicates")
    @Override
    public void listen(Handler<AsyncResult<JsonObject>> messageHandler) {
        this.vertx.executeBlocking(res -> {
        try{
            Future<TopicSubscriber> future = this.subscriberRef.get();
            if(future == null){
                Future<TopicSubscriber> newFuture = Future.future();
                //根据set结果判断subscriberRef是否已被并发更新
                if(this.subscriberRef.compareAndSet(null,newFuture) &&
                        this.cacheManager.cacheSubscriber(this)){
                    TopicSubscriber newSubscriber = session.createDurableSubscriber(topic, this.key);
                    newFuture.complete(newSubscriber);
                    //设置消息监听,将消息传给handler
                    newSubscriber.setMessageListener( message -> {
                        try {
                            String msg = ((ActiveMQTextMessage)message).getText();
                            messageHandler.handle(Future.succeededFuture(new JsonObject(msg)));
                        }catch (JMSException e){
                            messageHandler.handle(Future.failedFuture(e));
                        }
                    });
                }else{
                    //set失败,说明subscriberRef已被其他线程更新
                    messageHandler.handle(Future.failedFuture(new IllegalStateException("ActiveMQ主题订阅者监听已启动，请不要重复启动! key = " + this.key + ", queue = " + this.topic)));
                }
            }else{
                messageHandler.handle(Future.failedFuture(new IllegalStateException("ActiveMQ主题订阅者监听已启动，请不要重复启动! key = " + this.key + ", queue = " + this.topic)));
            }
        }catch (Exception e){
            messageHandler.handle(Future.failedFuture(e));
        }
        }, null);
    }

    /**
     * 关闭topic监听
     */
    @Override
    public void close() {
        Future<TopicSubscriber> future = this.subscriberRef.get();
        if(future != null){
            if(this.subscriberRef.compareAndSet(future, null)){
                try {
                    future.result().close();
                } catch (JMSException ignore) {
                    //ignore this exception
                } finally {
                    this.cacheManager.removeSubscriber(this.key);
                }
            }
        }
    }

    /**
     * 关闭topic监听
     */
    @Override
    public void close(Handler<AsyncResult<Void>> handler) {
        try {
            this.close();
            handler.handle(Future.succeededFuture());
        }catch (Exception e){
            handler.handle(Future.failedFuture(e));
        }
    }
}

File: src/main/kotlin/cn/booklish/vertx/activemq/client/subscriber/ActiveMQSubscriberImpl.kt
package cn.booklish.vertx.activemq.client.subscriber

import cn.booklish.vertx.activemq.client.cache.ActiveMQCacheManager
import cn.booklish.vertx.activemq.client.util.getUUID
import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.Vertx
import io.vertx.core.json.JsonObject
import org.apache.activemq.command.ActiveMQTextMessage
import java.util.concurrent.atomic.AtomicReference
import javax.jms.Session
import javax.jms.TopicSubscriber


class ActiveMQSubscriberImpl(override val key:String, private val vertx: Vertx, private val session: Session, destination: String) : ActiveMQSubscriber {

    private val topic = session.createTopic(destination)

    private var subscriberRef: AtomicReference<TopicSubscriber> = AtomicReference()

    /**
     * 开启订阅监听
     */
    override fun listen(messageHandler: Handler<AsyncResult<JsonObject>>) {
        vertx.executeBlocking(Handler<Future<JsonObject>>{
            try{
                val subscriber = this.subscriberRef.get()
                if(subscriber == null){
                    val newSubscriber = session.createDurableSubscriber(topic, getUUID())
                    //根据set结果判断subscriberRef是否已被并发更新
                    if(this.subscriberRef.compareAndSet(null,newSubscriber)){
                        //set成功
                        //加入缓存管理器
                        ActiveMQCacheManager.cacheSubscriber(this)
                        //设置消息监听,将消息传给handler
                        newSubscriber.setMessageListener {
                            messageHandler.handle(Future.succeededFuture(JsonObject((it as ActiveMQTextMessage).text)))
                        }
                    }else{
                        //set失败,说明subscriberRef已被其他线程更新,那么关闭新创建的newSubscriber释放资源
                        newSubscriber.close()
                        messageHandler.handle(Future.failedFuture(IllegalStateException("${this.subscriberRef.get()} had started, you should " +
                                "not call this method more than one time!")))
                    }
                }else{
                    messageHandler.handle(Future.failedFuture(IllegalStateException("$subscriber had started, you should " +
                            "not call this method more than one time!")))
                }
            }catch (e: Exception){
                messageHandler.handle(Future.failedFuture(e))
            }
        }, null)
    }

    /**
     * 关闭订阅
     */
    override fun close() {
        val subscriber = this.subscriberRef.get()
        if(subscriber != null){
            if(this.subscriberRef.compareAndSet(subscriber, null)){
                subscriber.close()
                ActiveMQCacheManager.removeSubscriber(this.key)
            }
        }
    }

    /**
     * 关闭订阅
     */
    override fun close(handler: Handler<AsyncResult<Void>>) {
        try {
            this.close()
            handler.handle(Future.succeededFuture())
        }catch (e: Exception){
            handler.handle(Future.failedFuture(e))
        }
    }

}